#### Riley Payung
#### CDS 292
#### 04/15/2020
#### Homework 8

#### Question 1

1 -> 2: 1

1 -> 3: 2

1 -> 4: 3

1 -> 5: 4

2 -> 1: 1

2 -> 3: 1

2 -> 4: 2

2 -> 5: 3

3 -> 1: 2

3 -> 2: 1

3 -> 4: 1

3 -> 5: 2

4 -> 1: 3

4 -> 2: 2

4 -> 3: 1

4 -> 5: 1

5 -> 1: 4

5 -> 2: 3

5 -> 3: 2

5 -> 4: 1

Our histogram would look like this:

|||1 2 3 4 5 6 7 8
   
1: - - - - - - - -

2: - - - - - -

3: - - - -

4: - -

5:

#### Question 2

As the general case, as you add more nodes, you increase all of the previous paths in the network by 1. example, if we were to have 6 nodes, we would have a histogram of the following data:

1. 9
2. 7
3. 5
4. 3
5. 1
6. 0

#### Question 6

.

.

.

.

.

.

.

.

.

.

.

.

.

#### Question 8

There are multiple, H(s|.) which is the histogram of the shortest paths from all origins to all other nodes, H(s|o) which is the shortest paths of length s from the specified 
origin to all other nodes; H(s|R=#), a histogram of the shortest paths of length s from a set of randomly chosen origins to all other consistently connected nodes, it is worthy
to note that the random origin could be in a cluster and that will be counted. All of these happen to be the types of searchs (1 to 3).

#### Question 12

.

.

.

.

.

.

.

.

.

#### Question 13

.

.

.

.

.

.

.

.

#### Question 14

.

.

.

.

.

.

.

#### Question 15

Large World of the previous question: 10^3 = 1000, 10^4 = 10000, 10^5 = 100000

Small World of the previous question: log(10^3) =  6.9, log(10^4) = 9.2, log(10^5) = 11.51

log(log(10^3)) = 1.93, log(log(10^4)) = 2.22, log(log(10^5)) = 2.44

.

#### Question 16

The network is a star network. The histogram would have a large number of s=1 paths along with a large number of s=2 paths; however, that is it, there will not
be any higher number of shells.

#### Question 21

In [43]:
import networkx as netx
import matplotlib.pyplot as plt

In [42]:
def BFS(G, o, dest=None):
    s = {};
    ts=0;
    w=list(G.nodes());
    s[o]=ts;
    w.remove(o);
    active=[o];
    while len(active) > 0:
        ts = ts+1
        newActive = [];
        for node in active:
            for neighbor in G.neighbors(node):
                if neighbor == dest:
                    s[neighbor] = ts
                    return (ts);
                if neighbor in w:
                    newActive.append(neighbor);
                    w.remove(neighbor);
                    s[neighbor] = ts;
            active = newActive;
        if (dest!=None):
            return -1;
        return s;
    
G = netx.Graph();
N = 100;
for i in range(1,N):
    G.add_node(i);

for i in range(1,N):
    for j in range(1,N):
        if ((i % 2 == 0) and (j == (i + 1)) and (not (G.has_edge(i,j) or G.has_edge(j,i)))):
            G.add_edge(i,j);

Hsn = {};
for node in G.nodes():
    SL=BFS(G,node);
    for s in SL.values():
        Hsn[s]=Hsn.get(s,0)+1;
print(Hsn)

Hsum = 0;
for i in Hsn.keys():
    Hsum=Hsum+Hsn[1];
print("Sum: ",Hsum)

{0: 99, 1: 98}
Sum:  196


It does apply since the n choose 2 for each cluster is size 2, so essentially 2 choose 2, which is 1, then we sum them all up and multiply by the number of shells. which would give us 196.